File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/resources/components.py
"""Components / Component Sets / Styles — Figma REST API coverage.

Tier: basic.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

if TYPE_CHECKING:
    from .._figma_client import FigmaClient


def _page(
    page_size: int | None = None,
    after: str | int | None = None,
    before: str | int | None = None,
) -> dict[str, Any]:
    params: dict[str, Any] = {}
    if page_size is not None:
        params["page_size"] = page_size
    if after is not None:
        params["after"] = after
    if before is not None:
        params["before"] = before
    return params


class ComponentsResource:
    def __init__(self, client: FigmaClient) -> None:
        self._client = client

    async def list_for_team(
        self,
        team_id: str,
        *,
        page_size: int | None = None,
        after: str | int | None = None,
        before: str | int | None = None,
    ) -> dict[str, Any]:
        """GET /v1/teams/:team_id/components — published components."""
        return await self._client.get(
            f"/v1/teams/{quote(team_id, safe='')}/components",
            params=_page(page_size, after, before),
        )

    async def list_for_file(self, file_key: str) -> dict[str, Any]:
        """GET /v1/files/:key/components — components defined in a file."""
        return await self._client.get(f"/v1/files/{quote(file_key, safe='')}/components")

    async def get(self, component_key: str) -> dict[str, Any]:
        """GET /v1/components/:key — fetch a published component by key."""
        return await self._client.get(f"/v1/components/{quote(component_key, safe='')}")

    async def list_component_sets_for_team(
        self,
        team_id: str,
        *,
        page_size: int | None = None,
        after: str | int | None = None,
        before: str | int | None = None,
    ) -> dict[str, Any]:
        """GET /v1/teams/:team_id/component_sets — published component sets."""
        return await self._client.get(
            f"/v1/teams/{quote(team_id, safe='')}/component_sets",
            params=_page(page_size, after, before),
        )

    async def list_component_sets_for_file(self, file_key: str) -> dict[str, Any]:
        """GET /v1/files/:key/component_sets — component sets inside a file."""
        return await self._client.get(
            f"/v1/files/{quote(file_key, safe='')}/component_sets"
        )

    async def get_component_set(self, set_key: str) -> dict[str, Any]:
        """GET /v1/component_sets/:key — single component set."""
        return await self._client.get(f"/v1/component_sets/{quote(set_key, safe='')}")

    async def list_styles_for_team(
        self,
        team_id: str,
        *,
        page_size: int | None = None,
        after: str | int | None = None,
        before: str | int | None = None,
    ) -> dict[str, Any]:
        """GET /v1/teams/:team_id/styles — published styles."""
        return await self._client.get(
            f"/v1/teams/{quote(team_id, safe='')}/styles",
            params=_page(page_size, after, before),
        )

    async def list_styles_for_file(self, file_key: str) -> dict[str, Any]:
        """GET /v1/files/:key/styles — styles inside a single file."""
        return await self._client.get(f"/v1/files/{quote(file_key, safe='')}/styles")

    async def get_style(self, style_key: str) -> dict[str, Any]:
        """GET /v1/styles/:key — single style."""
        return await self._client.get(f"/v1/styles/{quote(style_key, safe='')}")

File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/resources/components.ts
// @ts-nocheck
/**
 * Components / Component Sets / Styles — Figma REST API coverage.
 *
 * Covers the 3 library-object types under one resource. Each type
 * exposes the same 3 endpoints: list-by-team, list-by-file, get-by-key.
 *
 * Tier: basic (published components/styles are accessible to anyone
 * with the PAT; enterprise-only analytics are under LibraryAnalytics).
 */

import type { FigmaClient } from '../figma-client.js';

export interface TeamPageParams {
  page_size?: number;
  after?: string | number;
  before?: string | number;
}

function pageParams(p: TeamPageParams): Record<string, string | number> {
  const out: Record<string, string | number> = {};
  if (p.page_size !== undefined) out.page_size = p.page_size;
  if (p.after !== undefined) out.after = p.after;
  if (p.before !== undefined) out.before = p.before;
  return out;
}

export class ComponentsResource {
  constructor(private readonly client: FigmaClient) {}

  /** GET /v1/teams/:team_id/components — published components for a team. */
  async listForTeam(teamId: string, params: TeamPageParams = {}): Promise<unknown> {
    return this.client.get(`/v1/teams/${encodeURIComponent(teamId)}/components`, {
      params: pageParams(params),
    });
  }

  /** GET /v1/files/:key/components — components defined inside a single file. */
  async listForFile(fileKey: string): Promise<unknown> {
    return this.client.get(`/v1/files/${encodeURIComponent(fileKey)}/components`);
  }

  /** GET /v1/components/:key — fetch a single published component by key. */
  async get(componentKey: string): Promise<unknown> {
    return this.client.get(`/v1/components/${encodeURIComponent(componentKey)}`);
  }

  /** GET /v1/teams/:team_id/component_sets — published component sets. */
  async listComponentSetsForTeam(teamId: string, params: TeamPageParams = {}): Promise<unknown> {
    return this.client.get(`/v1/teams/${encodeURIComponent(teamId)}/component_sets`, {
      params: pageParams(params),
    });
  }

  /** GET /v1/files/:key/component_sets — component sets inside a file. */
  async listComponentSetsForFile(fileKey: string): Promise<unknown> {
    return this.client.get(`/v1/files/${encodeURIComponent(fileKey)}/component_sets`);
  }

  /** GET /v1/component_sets/:key — single component set. */
  async getComponentSet(setKey: string): Promise<unknown> {
    return this.client.get(`/v1/component_sets/${encodeURIComponent(setKey)}`);
  }

  /** GET /v1/teams/:team_id/styles — published styles (color/text/effect/grid). */
  async listStylesForTeam(teamId: string, params: TeamPageParams = {}): Promise<unknown> {
    return this.client.get(`/v1/teams/${encodeURIComponent(teamId)}/styles`, {
      params: pageParams(params),
    });
  }

  /** GET /v1/files/:key/styles — styles inside a single file. */
  async listStylesForFile(fileKey: string): Promise<unknown> {
    return this.client.get(`/v1/files/${encodeURIComponent(fileKey)}/styles`);
  }

  /** GET /v1/styles/:key — single style. */
  async getStyle(styleKey: string): Promise<unknown> {
    return this.client.get(`/v1/styles/${encodeURIComponent(styleKey)}`);
  }
}
